File: hooks/search-counter.py
#!/usr/bin/env python3
# SPDX-FileCopyrightText: 2026 Igor Santos
# SPDX-License-Identifier: MIT
# PreToolUse hook on Grep/Glob/Read: track exploration breadth. Nudge Claude
# toward the Explore subagent when the main session is fanning out across
# many files.
#
# Counting rules:
#   - Grep/Glob: each call = 1.
#   - Read: only the *first* time a unique absolute path is read this session
#     counts. Subsequent reads of the same file don't (they're often offset
#     follow-ups, which we want to encourage, not discourage).
#
# Emits additionalContext at thresholds 4, 8, 12. Past 12 it stays silent so
# it doesn't become spam: by then Claude has either delegated or chosen not to.

import os
import sys

sys.path.insert(0, os.path.join(os.path.dirname(os.path.abspath(__file__)), "lib"))
import common as c  # noqa: E402


def main() -> int:
    dir_ = c.session_dir()
    if not dir_:
        return 0

    tool = c.field(".tool_name")

    count_file = os.path.join(dir_, "search-count")
    seen_file = os.path.join(dir_, "seen-reads")
    tool_count_file = os.path.join(dir_, "tool-count")

    # Bump global tool counter (statusline reads this).
    c.incr_counter(tool_count_file)

    bump_search = False
    if tool in ("Grep", "Glob"):
        bump_search = True
    elif tool == "Read":
        path = c.field(".tool_input.file_path")
        if path:
            abs_path = c.abspath(path)
            if not _seen(seen_file, abs_path):
                try:
                    with open(seen_file, "a") as f:
                        f.write(abs_path + "\n")
                except Exception:
                    pass
                bump_search = True

    if not bump_search:
        return 0

    n = c.incr_counter(count_file)

    # Threshold nudges. Single, escalating message at each step.
    if n == 4:
        c.emit_pre_context(
            "PreToolUse",
            'Search/read count for this session has reached %d. If your remaining '
            'searches will fan across more than a couple more files, dispatch the '
            'Explore subagent now (Agent tool, subagent_type: "Explore"): its full '
            'search context stays in its window and only a digest comes back to '
            'yours. Keeps main context lean for the actual work.' % n,
        )
    elif n == 8:
        c.emit_pre_context(
            "PreToolUse",
            "Search/read count is now %d. You're deep in exploration, so strongly "
            "prefer dispatching the Explore subagent for the rest of this discovery "
            "work. Each additional Read here costs main-context tokens you won't "
            "recover." % n,
        )
    elif n == 12:
        c.emit_pre_context(
            "PreToolUse",
            "Search/read count is %d. Main context is now carrying significant "
            "exploration weight. Wrap up this discovery and continue in an Explore "
            "subagent, or summarize findings to yourself and consider /clear once "
            "the task is settled." % n,
        )
    return 0


def _seen(seen_file: str, abs_path: str) -> bool:
    """Whole-line exact match against seen-reads (mirrors grep -qxF)."""
    try:
        with open(seen_file) as f:
            for line in f:
                if line.rstrip("\n") == abs_path:
                    return True
    except Exception:
        return False
    return False


if __name__ == "__main__":
    sys.exit(main())

File: src/hooks/search_counter.rs
// SPDX-FileCopyrightText: 2026 Igor Santos
// SPDX-License-Identifier: MIT

//! Ports hooks/search-counter.py: a PreToolUse hook on Grep/Glob/Read that
//! tracks exploration breadth and nudges Claude toward the Explore subagent
//! once the main session fans out across many files.
//!
//! Counting rules:
//!   - Grep/Glob: each call counts 1.
//!   - Read: only the first time a unique absolute path is read this session
//!     counts. Subsequent reads of the same file don't, since those are
//!     often offset follow-ups that should be encouraged, not discouraged.
//!
//! Emits additionalContext at thresholds 4, 8 and 12. Past 12 it stays
//! silent so it doesn't become spam: by then Claude has either delegated or
//! chosen not to.

use crate::common::payload::Payload;
use crate::common::{abspath, atomic_append, emit_pre_context, incr_counter, session_dir};
use std::fs;

pub fn run(payload: &Payload) {
    let dir = session_dir(payload);
    if dir.is_empty() {
        return;
    }

    let tool = payload.field(".tool_name");

    let count_file = format!("{dir}/search-count");
    let seen_file = format!("{dir}/seen-reads");
    let tool_count_file = format!("{dir}/tool-count");

    // Bump global tool counter (statusline reads this).
    incr_counter(&tool_count_file);

    let mut bump_search = false;
    if tool == "Grep" || tool == "Glob" {
        bump_search = true;
    } else if tool == "Read" {
        let path = payload.field(".tool_input.file_path");
        if !path.is_empty() {
            let abs_path = abspath(&path);
            if !seen(&seen_file, &abs_path) {
                // hooks/search-counter.py appends to seen_file with a plain
                // unlocked `open(..., "a")`. The bytes written here are
                // identical; only the synchronisation differs, a deliberate
                // difference rather than a divergence in behaviour.
                atomic_append(&seen_file, &abs_path);
                bump_search = true;
            }
        }
    }

    if !bump_search {
        return;
    }

    let n = incr_counter(&count_file);
    nudge(n);
}

/// Escalating nudge toward the Explore subagent at counts 4, 8 and 12. Any
/// other count, including everything past 12, stays silent so the nudge
/// never becomes spam.
fn nudge(n: i64) {
    match n {
        4 => emit_pre_context(
            "PreToolUse",
            &format!(
                "Search/read count for this session has reached {n}. If your remaining \
                 searches will fan across more than a couple more files, dispatch the \
                 Explore subagent now (Agent tool, subagent_type: \"Explore\"): its full \
                 search context stays in its window and only a digest comes back to \
                 yours. Keeps main context lean for the actual work."
            ),
        ),
        8 => emit_pre_context(
            "PreToolUse",
            &format!(
                "Search/read count is now {n}. You're deep in exploration, so strongly \
                 prefer dispatching the Explore subagent for the rest of this discovery \
                 work. Each additional Read here costs main-context tokens you won't \
                 recover."
            ),
        ),
        12 => emit_pre_context(
            "PreToolUse",
            &format!(
                "Search/read count is {n}. Main context is now carrying significant \
                 exploration weight. Wrap up this discovery and continue in an Explore \
                 subagent, or summarize findings to yourself and consider /clear once \
                 the task is settled."
            ),
        ),
        _ => {}
    }
}

/// Whole-line exact match against `seen_file`, mirroring `grep -qxF`. A
/// missing or unreadable file is treated as "not seen". Never panics.
fn seen(seen_file: &str, abs_path: &str) -> bool {
    let Ok(contents) = fs::read_to_string(seen_file) else {
        return false;
    };
    contents.lines().any(|line| line == abs_path)
}
